### PR TITLE
Redesign Tabs screen.

### PR DESCRIPTION
**Pull request by dbrant**
### What does this do?
* Remove dependency on an unmaintained, abandoned third-party library.
* Use a purely standard RecyclerView to accomplish displaying tabs, complete with swiping, animations, etc.
* Make the Tabs list more modern, and more consistent with current expectations.

Before / After:

image  image

**Phabricator:**
https://phabricator.wikimedia.org/T339295

### CODE REVIEW

File: app/src/main/java/org/wikipedia/page/tabs/SwipeableTabTouchHelperCallback.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        fun isSwipeable(): Boolean
+    }
+
+    private val swipeBackgroundPaint = Paint()
+    private val itemBackgroundPaint = Paint()
+    var swipeableEnabled = false
+
+    init {
+        swipeBackgroundPaint.style = Paint.Style.FILL
+        swipeBackgroundPaint.color = getThemedColor(context, R.attr.background_color)
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/page/tabs/SwipeableTabTouchHelperCallback.kt:25]: Can we use `ResourceUtil.getThemedColor` here?

File: app/src/main/res/layout/item_tab_contents.xml [hunk 2]
```diff
@@ -1,40 +1,85 @@
...
+            android:id="@+id/tabContainer"
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            android:background="?attr/selectableItemBackground">
+
+            <ImageView
+                android:id="@+id/tabCloseButton"
+                android:layout_width="24dp"
+                android:layout_height="24dp"
+                android:padding="4dp"
```
Comment (cooltey) [inline app/src/main/res/layout/item_tab_contents.xml:26]: Would it be possible to set the `width` and `height` to `48dp` with some extra padding setup in order to follow the accessibility guidelines?
```

                android:layout_width="48dp"
                android:layout_height="48dp"
                android:paddingBottom="26dp"
                android:paddingStart="26dp"
                android:paddingEnd="4dp"
                android:paddingTop="4dp"
```

File: app/src/main/java/org/wikipedia/page/tabs/TabActivity.kt [hunk 3]
```diff
@@ -229,78 +146,28 @@ class TabActivity : BaseActivity() {
...
-        override fun onTabAdded(tabSwitcher: TabSwitcher, index: Int, tab: de.mrapp.android.tabswitcher.Tab, animation: Animation) {
-            binding.tabCountsView.updateTabCount(false)
-            tabUpdatedTimeMillis = System.currentTimeMillis()
-        }
-
-        override fun onTabRemoved(tabSwitcher: TabSwitcher, index: Int, tab: de.mrapp.android.tabswitcher.Tab, animation: Animation) {
-            if (app.tabList.isNotEmpty() && index < app.tabList.size) {
-                val tabIndex = app.tabList.size - index - 1
-                val appTab = app.tabList.removeAt(tabIndex)
+                binding.tabRecyclerView.adapter?.notifyDataSetChanged()
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/page/tabs/TabActivity.kt:167]: Would it be a good idea to use `binding.tabRecyclerView.adapter?.notifyItemRangeChanged(0, app.tabList.size)` for a smoother animation?

File: app/src/main/java/org/wikipedia/page/tabs/TabActivity.kt [hunk 4]
```diff
@@ -328,35 +195,83 @@ class TabActivity : BaseActivity() {
...
+        override fun isSwipeable(): Boolean {
+            return true
+        }
+
+        private fun doCloseTab() {
+            val adapterPosition = bindingAdapterPosition
+            val index = adapterPositionToTabIndex(adapterPosition)
+            val appTab = app.tabList.removeAt(index)
+            binding.tabCountsView.updateTabCount(false)
+            bindingAdapter?.notifyItemRemoved(adapterPosition)
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/page/tabs/TabActivity.kt:250]: We user removes the first tab from the list (the one with a blue border), the second tab moves to the first position, but does not get the border set.

File: app/src/main/java/org/wikipedia/views/AutoFitRecyclerView.kt [hunk 5]
```diff
@@ -4,30 +4,40 @@ import android.content.Context
...
     interface Callback {
         fun onColumns(columns: Int)
     }
 
-    private val minColumnWidth = context.obtainStyledAttributes(attrs, R.styleable.AutoFitRecyclerView, defStyleAttr, 0).use {
-        it.getDimensionPixelSize(R.styleable.AutoFitRecyclerView_minColumnWidth, 0)
-    }
-    @AndroidIntRange(from = MIN_COLUMNS.toLong())
-    var columns = MIN_COLUMNS
+    protected var recyclerLayoutManager: StaggeredGridLayoutManager
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/views/AutoFitRecyclerView.kt:19]: Not sure but should this be `private`?

**Review by cooltey — CHANGES_REQUESTED** [review]
Great job! Just have some comments 👍

File: app/src/main/java/org/wikipedia/page/tabs/TabActivity.kt [hunk 6]
```diff
@@ -57,78 +47,13 @@ class TabActivity : BaseActivity() {
...
-        binding.tabSwitcher.showSwitcher()
-
-        binding.tabSwitcher.addCloseTabListener { tabSwitcher, tab ->
-            tabSwitcher.removeTab(tab, SwipeAnimation.Builder()
-                .setDuration(0)
-                .setRelocateAnimationDuration(100)
-                .setInterpolator(null).create())
-            false
-        }
+        binding.tabRecyclerView.itemAnimator = DefaultItemAnimator()
```
[the remark this answers is not in the document]
Comment (dbrant) [reply app/src/main/java/org/wikipedia/page/tabs/TabActivity.kt]: Nope!

File: app/src/main/java/org/wikipedia/page/tabs/SwipeableTabTouchHelperCallback.kt (hunk 1 above)
Comment (dbrant) [reply app/src/main/java/org/wikipedia/page/tabs/SwipeableTabTouchHelperCallback.kt]: yep!

File: app/src/main/java/org/wikipedia/page/tabs/TabActivity.kt (hunk 3 above)
Comment (dbrant) [reply app/src/main/java/org/wikipedia/page/tabs/TabActivity.kt]: yep

File: app/src/main/java/org/wikipedia/views/AutoFitRecyclerView.kt (hunk 5 above)
Comment (dbrant) [reply app/src/main/java/org/wikipedia/views/AutoFitRecyclerView.kt]: yep

File: app/src/main/java/org/wikipedia/page/tabs/TabActivity.kt (hunk 4 above)
Comment (dbrant) [reply app/src/main/java/org/wikipedia/page/tabs/TabActivity.kt:251]: Done!

File: app/src/main/res/layout/item_tab_contents.xml (hunk 2 above)
Comment (dbrant) [reply app/src/main/res/layout/item_tab_contents.xml]: The current design is consistent with Google's own design in the Chrome app, but I do agree with your suggestion, with a slight tweak.

**Pull request merged**